### DISCUSSION TITLE
LDAP: select alternative external_identifier

### DISCUSSION THREAD

**Discussion by takuy** [directus/directus#10165] [Feature Requests]
For the LDAP auth driver, using the user DN as the **external_identifier** just isn't suitable in my environment. It's subject to change. We have immutable IDs affiliated with users, so I think the ability to use an arbitrary attribute (or just USER_ATTRIBUTE) would be very use. I don't think this is possible at the moment.

**Comment by aidenfoxx** [comment]
This seems like a very weird use case? Why would you change a users DN/CN?

**Reply by takuy** [reply]
The DN changes based on their user object's location in the directory. It's not always the same path. In our AD environment, unfortunately, this means they could be moved around to various OU depending on their department or based on a script that moves accounts based on status in the organization. 

Our CNs also contain user-entered first and last names, which are also subject to change, and are also included as part of the DN. Not sure who decided this but way out of my control. 

The user's username (in AD, usually sAMAccountName or userPrincipalName) would never change. Nor would something like an `employeeId`, at least not in my case.

**Reply by aidenfoxx** [reply]
Using "sAMAccountName" or "userPrincipalName" as an identifier is fine, and can be used for authentication. But for simplicity sake it's good to keep the users full DN around for querying. We do this, for example, when the users auth token is refreshed. We use their `external_identifier` to check that their account status is still valid.

Obviously this could also be achieved with an alternate identifier like "sAMAccountName", but it'll take some work to implement. But I'd welcome the PR if you feel up to the challenge!

**Reply by aidenfoxx** [reply]
Otherwise you could rely on "sAMAccountName" or "userPrincipalName" for authentication, and the user would just have to deal with getting a new account if they change department. Shouldn't be too much of an issue since roles are assigned based on LDAP groups, so their permissions wouldn't change.

**Reply by takuy** [reply]
Might try to implement when things calm down (lol). My main problem with the DN approach at the moment is that the Directus the auth seems to fail when the DN contains a parenthesis (an example of my cn/dn struggle). 

```log

20:01:39 🔍 [1.123ms] select `id` from `directus_users` where LOWER(`external_identifier`) = ? limit ? [cn=lname\, fname (username),ou=snip,ou=snop,dc=com, 1]
20:01:39 🚨 illegal unescaped char: ( 
Error: illegal unescaped char: (
    at escapeValue (/directus/node_modules/ldap-filter/lib/index.js:49:13)
    at parseExpr (/directus/node_modules/ldap-filter/lib/index.js:167:16)
    at parseFilter (/directus/node_modules/ldap-filter/lib/index.js:228:14)
    at Object.parse (/directus/node_modules/ldap-filter/lib/index.js:252:18)
    at Object.parseString (/directus/node_modules/ldapjs/lib/filters/index.js:179:27)
    at Client.search (/directus/node_modules/ldapjs/lib/client/client.js:571:30)
    at /directus/node_modules/directus/dist/auth/drivers/ldap.js:152:29
    at new Promise (<anonymous>)
    at LDAPAuthDriver.fetchUserGroups (/directus/node_modules/directus/dist/auth/drivers/ldap.js:149:16)
    at LDAPAuthDriver.getUserID (/directus/node_modules/directus/dist/auth/drivers/ldap.js:197:39)
20:01:39 ✨ request errored POST 500 /auth/login/SSO 24ms
```

**Reply by aidenfoxx** [reply]
Oh damn. I'm not sure if that's on `ldapjs`/`ldap-filter` or us. Since we just send the `dn` we get back from `ldapjs` I'd wager it's a problem on their side. Not sure if we could mitigate it in any way. 🤔

**Reply by aidenfoxx** [reply]
This should be resolved with https://github.com/directus/directus/pull/10297

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
